simul/platform/mininet.go

```diff
@@ -0,0 +1,372 @@
+	m.config = simulConfig.Config
+	log.Lvl3("Saving configuration")
+	simulConfig.Save(m.buildDir)
+
+	// Verify the installation is correct
+	gw := m.HostIPs[0]
+	log.Lvl2("Verifying configuration on", gw)
+	out, err := exec.Command("ssh", "root@"+gw, "which mn").Output()
+	if err != nil || !strings.HasSuffix(string(out), "mn\n") {
+		log.Error("While trying to connect to", gw, err)
+		log.Fatal("Please verify installation of mininet or run\n" +
+			"./platforms/mininet/setup_iccluster.sh")
+	}
+
+	// Copy our script
+	err = Copy(m.buildDir, m.mininetDir+"/start.py")
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	// Copy everything over to MiniNet
+	log.Lvl1("Copying over to", m.Login, "@", m.External)
+	err = Rsync(m.Login, m.External, m.buildDir+"/", "mininet_run/")
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Lvl2("Done copying")
+
+	return nil
+}
+
+// Start connects to the first of the remote servers to start the simulation.
+func (m *MiniNet) Start(args ...string) error {
+	// setup port forwarding for viewing log server
+	m.started = true
+	// Remote tunneling : the sink port is used both for the sink and for the
+	// proxy => the proxy redirects packets to the same port the sink is
+	// listening.
+	// -n = stdout == /Dev/null, -N => no command stream, -T => no tty
+	var exCmd *exec.Cmd
+	redirection := fmt.Sprintf("*:%d:%s:%d", m.MonitorPort, m.ProxyAddress, m.MonitorPort)
+	login := fmt.Sprintf("%s@%s", m.Login, m.External)
+	cmd := []string{"-nNTf", "-o", "StrictHostKeyChecking=no", "-o", "ExitOnForwardFailure=yes", "-R",
+		redirection, login}
+	exCmd = exec.Command("ssh", cmd...)
+	if err := exCmd.Start(); err != nil {
+		log.Fatal("Failed to start the ssh port forwarding:", err)
+	}
+	if err := exCmd.Wait(); err != nil {
+		log.Fatal("ssh port forwarding exited in failure:", err)
+	}
+	go func() {
+		config := strings.Split(m.config, "\n")
+		sort.Strings(config)
+		log.Lvlf1("Starting simulation %s over mininet", strings.Join(config, " :: "))
+		err := SSHRunStdout(m.Login, m.External, "cd mininet_run; ./start.py list go")
+		if err != nil {
+			log.Lvl3(err)
+		}
+		m.sshMininet <- "finished"
+	}()
+
+	return nil
+}
+
+// Wait blocks on the channel till the main-process finishes.
+func (m *MiniNet) Wait() error {
+	wait := m.CloseWait
+	if wait == 0 {
+		wait = 600
+	}
+	if m.started {
+		log.Lvl3("Simulation is started")
+		select {
+		case msg := <-m.sshMininet:
+			if msg == "finished" {
+				log.Lvl3("Received finished-message, not killing users")
+				return nil
+			}
+			log.Lvl1("Received out-of-line message", msg)
+		case <-time.After(time.Second * time.Duration(wait)):
+			log.Lvl1("Quitting after ", wait/60,
+				" minutes of waiting")
+			m.started = false
+		}
+		m.started = false
+	}
+	return nil
+}
+
+// Returns the servers to use for mininet.
+func (m *MiniNet) parseServers() error {
+	hosts, err := ioutil.ReadFile(path.Join(m.mininetDir, "server_list"))
+	if err != nil {
```

Comment:
I'd rather avoid user-inputs. Deterlab is special, because you don't change the configuration. Whereas mininet (at least with iccluster) will need a new set of servers every simulation... But a better error-message is the least I can do.